### PR TITLE
feat: Added module virtual-network

### PR DESCRIPTION
**Pull request by elbatane**
## Description

Added module virtual-network - converted from CARML.

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

Enable GitHub Worksflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) as a secret in your forked repository called `PAT`.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

## Updating an existing module

- [ ] This is a bug fix:
  - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
  - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
- [ ] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.
- [ ] I have read the [Updating an existing module](https://github.com/Azure/bicep-registry-modules/blob/main/CONTRIBUTING.md#updating-an-existing-module) section in the contributing guide and updated the `version.json` file properly:
  - [ ] The PR contains backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`.
  - [ ] The PR contains backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] The PR contains breaking changes, and I have bumped the MAJOR version in `version.json`.
- [ ] I have updated the examples in README with the latest module version number.

### CODE REVIEW

File: avm/res/network/virtual-network/tests/e2e/vnetPeering/main.test.bicep [hunk 1]
```diff
@@ -0,0 +1,93 @@
+targetScope = 'subscription'
+
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/tests/e2e/vnetPeering/main.test.bicep:2]:
```suggestion

metadata name = 'Deploying a bi-directional peering'
metadata description = 'This instance deploys the module with both an inbound and outbound peering.'

```

Comment (AlexanderSehr) [reply avm/res/network/virtual-network/tests/e2e/vnetPeering/main.test.bicep:2]: > **Note:** This requires a regeneration of the top-level readme

File: avm/res/network/virtual-network/tests/e2e/waf-aligned/main.test.bicep [hunk 2]
```diff
@@ -0,0 +1,167 @@
+targetScope = 'subscription'
+
+metadata name = 'Using large parameter set'
+metadata description = 'This instance deploys the module with most of its features enabled.'
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/tests/e2e/waf-aligned/main.test.bicep:4]:
```suggestion
metadata name = 'WAF-aligned'
metadata description = 'This instance deploys the module in alignment with the best-practices of the Well-Architected Framework.'
```

Comment (AlexanderSehr) [reply avm/res/network/virtual-network/tests/e2e/waf-aligned/main.test.bicep:4]: > **Note:** This requires the top-level readme to be regenerated

File: avm/res/network/virtual-network/tests/e2e/waf-aligned/main.test.bicep [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      kind: 'CanNotDelete'
+      name: 'myCustomLockName'
+    }
+    roleAssignments: [
+      {
+        roleDefinitionIdOrName: 'Reader'
+        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+        principalType: 'ServicePrincipal'
+      }
+    ]
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/tests/e2e/waf-aligned/main.test.bicep:103]:
```suggestion
```

Not required for WAF-alignment. 
> **Note:** This requires the top-level readme to be regenerated

File: avm/res/network/virtual-network/subnet/main.bicep [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+param name string
+
+@description('Conditional. The name of the parent virtual network. Required if the template is used in a standalone deployment.')
+param virtualNetworkName string
+
+@description('Required. The address prefix for the subnet.')
+param addressPrefix string
+
+@description('Optional. The resource ID of the network security group to assign to the subnet.')
+param networkSecurityGroupId string = ''
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/subnet/main.bicep:15]:
```suggestion
param networkSecurityGroupResourceId string = ''
```

File: avm/res/network/virtual-network/subnet/main.bicep [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+@description('Optional. The resource ID of the route table to assign to the subnet.')
+param routeTableId string = ''
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/subnet/main.bicep:18]:
```suggestion
param routeTableResourceId string = ''
```

File: avm/res/network/virtual-network/subnet/main.bicep [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+@description('Optional. The service endpoints to enable on the subnet.')
+param serviceEndpoints array = []
+
+@description('Optional. The delegations to enable on the subnet.')
+param delegations array = []
+
+@description('Optional. The resource ID of the NAT Gateway to use for the subnet.')
+param natGatewayId string = ''
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/subnet/main.bicep:27]:
```suggestion
param natGatewayResourceId string = ''
```

File: avm/res/network/virtual-network/subnet/main.bicep [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    addressPrefix: addressPrefix
+    networkSecurityGroup: !empty(networkSecurityGroupId) ? {
+      id: networkSecurityGroupId
+    } : null
+    routeTable: !empty(routeTableId) ? {
+      id: routeTableId
+    } : null
+    natGateway: !empty(natGatewayId) ? {
+      id: natGatewayId
+    } : null
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/subnet/main.bicep:86]:
```suggestion
    networkSecurityGroup: !empty(networkSecurityGroupResourceId) ? {
      id: networkSecurityGroupResourceId
    } : null
    routeTable: !empty(routeTableResourceId) ? {
      id: routeTableResourceId
    } : null
    natGateway: !empty(natGatewayResourceId) ? {
      id: natGatewayResourceId
    } : null
```

File: avm/res/network/virtual-network/subnet/main.bicep [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+@description('The resource ID of the virtual network peering.')
+output resourceId string = subnet.id
+
+@description('The address prefix for the subnet.')
+output subnetAddressPrefix string = subnet.properties.addressPrefix
+
+@description('List of address prefixes for the subnet.')
+output subnetAddressPrefixes array = !empty(addressPrefixes) ? subnet.properties.addressPrefixes : []
+// =============== //
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/subnet/main.bicep:126]:
```suggestion

// =============== //
```

File: avm/res/network/virtual-network/main.bicep [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+param diagnosticSettings diagnosticSettingType
+
+@description('Optional. The lock settings of the service.')
+param lock lockType
+
+@description('Optional. Array of role assignment objects that contain the \'roleDefinitionIdOrName\' and \'principalId\' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'.')
+param roleAssignments roleAssignmentType
+
+@description('Optional. Tags of the resource.')
+param tags object = {}
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/main.bicep:50]:
```suggestion
param tags object?
```

File: avm/res/network/virtual-network/main.bicep [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+param addressPrefixes array
+
+@description('Optional. An Array of subnets to deploy to the Virtual Network.')
+param subnets array = []
+
+@description('Optional. DNS Servers associated to the Virtual Network.')
+param dnsServers array = []
+
+@description('Optional. Resource ID of the DDoS protection plan to assign the VNET to. If it\'s left blank, DDoS protection will not be configured. If it\'s provided, the VNET created by this template will be attached to the referenced DDoS protection plan. The DDoS protection plan can exist in the same or in a different subscription.')
+param ddosProtectionPlanId string = ''
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/main.bicep:21]:
```suggestion
param ddosProtectionPlanResourceId string = ''
```

Comment (AlexanderSehr) [reply avm/res/network/virtual-network/main.bicep:21]: Also please update the other references of this parameter

File: avm/res/network/virtual-network/main.bicep [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+
+type roleAssignmentType = {
+  @description('Required. The name of the role to assign. If it cannot be found you can specify the role definition ID instead.')
+  roleDefinitionIdOrName: string
+
+  @description('Required. The principal ID of the principal (user/group/identity) to assign the role to.')
+  principalId: string
+
+  @description('Optional. The principal type of the assigned principal ID.')
+  principalType: ('ServicePrincipal' | 'Group' | 'User' | 'ForeignGroup' | 'Device' | null)?
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/main.bicep:283]:
```suggestion
  principalType: ('ServicePrincipal' | 'Group' | 'User' | 'ForeignGroup' | 'Device')?
```

File: avm/res/network/virtual-network/main.bicep [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+  }[]?
+
+  @description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
+  metricCategories: {
+    @description('Required. Name of a Diagnostic Metric category for a resource type this setting is applied to. Set to \'AllMetrics\' to collect all metrics.')
+    category: string
+  }[]?
+
+  @description('Optional. A string indicating whether the export to Log Analytics should use the default destination type, i.e. AzureDiagnostics, or use a destination type.')
+  logAnalyticsDestinationType: ('Dedicated' | 'AzureDiagnostics' | null)?
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/main.bicep:318]:
```suggestion
  logAnalyticsDestinationType: ('Dedicated' | 'AzureDiagnostics')?
```

File: avm/res/network/virtual-network/subnet/main.bicep [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+type roleAssignmentType = {
+  @description('Required. The name of the role to assign. If it cannot be found you can specify the role definition ID instead.')
+  roleDefinitionIdOrName: string
+
+  @description('Required. The principal ID of the principal (user/group/identity) to assign the role to.')
+  principalId: string
+
+  @description('Optional. The principal type of the assigned principal ID.')
+  principalType: ('ServicePrincipal' | 'Group' | 'User' | 'ForeignGroup' | 'Device' | null)?
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/subnet/main.bicep:138]:
```suggestion
  principalType: ('ServicePrincipal' | 'Group' | 'User' | 'ForeignGroup' | 'Device')?
```

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
Please fetch latest upstream and regenerate templates & readme's using the `Set-AVMModule` script :)

**Review by skydiver1980 — APPROVED** [review]

File: avm/res/network/virtual-network/tests/e2e/defaults/main.test.bicep [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+  params: {
+    name: '${namePrefix}${serviceShort}001'
+    location: location
+    addressPrefixes: [
+      '10.0.0.0/16'
+    ]
+    // Workaround for PSRule
+    lock: {}
+    diagnosticSettings: []
+    roleAssignments: []
```
Comment (AlexanderSehr) [inline avm/res/network/virtual-network/tests/e2e/defaults/main.test.bicep:51]:
```suggestion
```

This was finally fixed recently :) Hence we can remove it again - but need to refresh the readme

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**